hypertrace-gradle-publish-maven-central-plugin/src/main/java/org/hypertrace/gradle/publishing/PublishMavenCentralPlugin.java

```diff
@@ -0,0 +1,196 @@
+  private static final String PROPERTY_SIGNING_PASSWORD = "signingPassword";
+  private static final String PROPERTY_OSSRH_USERNAME = "ossrhUsername";
+  private static final String PROPERTY_OSSRH_PASSWORD = "ossrhPassword";
+
+  private Project project;
+  private HypertracePublishMavenCentralExtension extension;
+
+  @Override
+  public void apply(@Nonnull Project target) {
+    project = target;
+    this.extension =
+      project.getExtensions().create(EXTENSION_NAME, HypertracePublishMavenCentralExtension.class);
+    this.applyMavenPublish();
+    this.applySigning();
+    this.applyWithJavadocJar();
+    this.applyWithSourcesJar();
+    this.addPublishRepository();
+    this.addPublications();
+  }
+
+  private void applySigning() {
+    project.getPluginManager()
+      .apply(SigningPlugin.class);
+  }
+
+  private void applyMavenPublish() {
+    project.getPluginManager()
+      .apply(MavenPublishPlugin.class);
+  }
+
+  private void applyWithJavadocJar() {
+    getJavaPluginExtension().withJavadocJar();
+    project.getTasks().withType(Javadoc.class).configureEach(javadoc -> {
+      StandardJavadocDocletOptions options = (StandardJavadocDocletOptions) javadoc.getOptions();
+      if (JavaVersion.current().isJava9Compatible()) {
+        options.addBooleanOption("html5", true);
+      }
+    });
+  }
+
+  private void applyWithSourcesJar() {
+    getJavaPluginExtension().withSourcesJar();
+  }
+
+  private void addPublishRepository() {
+    String url;
+    String user = getPropertyOrThrow(project, PROPERTY_OSSRH_USERNAME);
+    String password = getPropertyOrThrow(project, PROPERTY_OSSRH_PASSWORD);
+
+    if (project.getVersion().toString().endsWith("SNAPSHOT")) {
+      url = "https://s01.oss.sonatype.org/content/repositories/snapshots/";
+    } else {
+      url = "https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/";
+    }
+
+    getPublishingExtension().repositories(artifactRepositories ->
+      artifactRepositories.maven(mavenArtifactRepository -> {
+        mavenArtifactRepository.setUrl(url);
+        mavenArtifactRepository.credentials(passwordCredentials -> {
+          passwordCredentials.setUsername(user);
+          passwordCredentials.setPassword(password);
+        });
+      })
+    );
+  }
+
+  private void addPublications() {
+    project.afterEvaluate(unused -> {
+      getPublishingExtension().publications(this::addJavaLibraryPublicationWhenApplied);
+
+      // sign after publication is added
+      addSigning();
+    });
+  }
+
+  private void addJavaLibraryPublicationWhenApplied(PublicationContainer publications) {
+    project.getPluginManager()
+      .withPlugin("java-library", appliedPlugin -> {
+        publications.create("javaLibrary", MavenPublication.class, publication -> {
+          // from
+          publication.from(project.getComponents().getByName("java"));
+
+          // versionMapping
+          publication.versionMapping(versionMappingStrategy -> {
```

Comment:
Fancy - do we need the version mapping? My understanding is this publishes the resolution result instead of the declared versions, right? This isn't something we do with our publications today (since a consumer resolves versions themselves, anyway), so I'd suggest keeping it simple unless we have a need.

Reply:
removed `versionMapping`.